src/ProxyServer/ProxyServer.swift

```diff
@@ -29,6 +29,8 @@ open class ProxyServer: NSObject, TunnelDelegate {
     open var observer: Observer<ProxyServerEvent>?
 
     var tunnels: TunnelArray = []
+    
+    private let accessQueue = DispatchQueue(label: "NEKit.ProxyServer.TunnelArrayAccess", attributes: .concurrent)
```

Comment:
You are right.  I wrote some debug logs:

```Swift
func didAcceptNewSocket(_ socket: ProxySocket) {
        DDLogDebug("=========== accept socket on queue: \(QueueFactory.onQueue())")
        
        ...
        
        
func tunnelDidClose(_ tunnel: Tunnel) {
        DDLogDebug("=========== tunnel did close on queue: \(QueueFactory.onQueue())")
```

There is no race condition.  

I will review it later, `Swift.Array`'s behavior is a little strange.